beacon-chain/blockchain/process_block.go

```diff
@@ -226,6 +227,13 @@ func (s *Service) onBlockBatch(ctx context.Context, blks []*ethpb.SignedBeaconBl
 		if err != nil {
 			return nil, nil, nil, err
 		}
+		r, err := stateutil.BlockRoot(b.Block)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		if err := s.stateGen.SaveState(ctx, r, preState); err != nil {
```

Comment:
the issue here, is that all signatures are not verified yet so saving states here might lead to bad states being persisted in our hot state cache. Also this state is not copied anywhere in that method, so there is the possibility of our cache having incorrect states for certain roots. In the loop, all hot states saved will refer to the same state pointer.

Reply:
This is ok. During inital syncing, we process blocks in the form of linked list. Every previous block is parent of new block. This implies:
1.) Copy is not necessary
2.) Hot state is not used. Even if it's used when one block fails, the whole batch would have failed anyway which invalidates the cache.
3.) We only cache epoch boundary cache states. When we retrieve the state in the epoch boundary state cache. The root in which the state we retrieve is finalized. We can be sure of it's correctness in that regard.
https://github.com/prysmaticlabs/prysm/blob/master/beacon-chain/state/stategen/migrate.go#L109